editor/store/reducer.js

```diff
@@ -582,6 +582,64 @@ export function blocksMode( state = {}, action ) {
  * Reducer returning the block insertion point visibility, a boolean value
  * reflecting whether the insertion point should be shown.
  *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Dispatched action
+ * @return {Object}        Updated state
+ */
+export function currentlyAutosaving( state = false, action ) {
+	switch ( action.type ) {
+		case 'DOING_AUTOSAVE':
+			const { isAutosaving } = action;
+			return isAutosaving;
+	}
+
+	return state;
+}
+
+/**
+ * Reducer returning the currently autosaving status.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Dispatched action
+ * @return {Object}        Updated state
+ */
+export const autosave = withChangeDetection( function( state = { message: '', autosave: false }, action ) {
+	switch ( action.type ) {
+		case 'RESET_AUTOSAVE':
+			return {
+				...state,
+				autosave: action.autosave,
+			};
+		case 'UPDATE_AUTOSAVE_STATUS_MESSAGE':
+			return {
+				...state,
+				message: action.message,
+			};
+	}
+
+	return state;
+}, { resetTypes: [ 'UPDATE_AUTOSAVE_STATUS_MESSAGE' ] } );
```

Comment:
It's not entirely obvious that `UPDATE_AUTOSAVE_STATUS_MESSAGE` is better represented as the action type which occurs as the result of a successful autosave. Maybe the status message should be updated as a separate (side-effect?) action?

Reply:
I'm not sure I understand what you mean "update as a separate action", how would this be implemented?